x-pack/lib/filters/geoip/database_metadata.rb

```diff
@@ -11,61 +11,78 @@ module LogStash module Filters module Geoip class DatabaseMetadata
   include LogStash::Util::Loggable
   include LogStash::Filters::Geoip::Util
 
-  def initialize(database_type)
-    @metadata_path = get_file_path("metadata.csv")
-    @database_type = database_type
+  def initialize
+    @metadata_path = ::File.join(get_data_dir_path, "metadata.csv")
   end
 
   public
 
-  # csv format: database_type, update_at, gz_md5, md5, filename
-  def save_timestamp(database_path)
-    metadata = get_metadata(false)
-    metadata << [@database_type, Time.now.to_i, md5(get_gz_name(database_path)), md5(database_path),
-                 ::File.basename(database_path)]
+  # csv format: database_type, update_at, gz_md5, dirname, is_eula
+  def save_metadata(database_type, dirname, is_eula)
+    metadata = get_metadata(database_type, false)
+    metadata << [database_type, Time.now.to_i, md5(get_gz_path(database_type, dirname)),
+                 dirname, is_eula]
+    update(metadata)
+  end
+
+  def update_timestamp(database_type)
+    metadata = get_all.map do |row|
+      row[Column::UPDATE_AT] = Time.now.to_i if row[Column::DATABASE_TYPE].eql?(database_type)
+      row
+    end
+    update(metadata)
+  end
 
+  def reset_md5(database_type)
+    metadata = get_all.map do |row|
+      row[Column::GZ_MD5] = ""  if row[Column::DATABASE_TYPE].eql?(database_type)
+      row
+    end
+    update(metadata)
+  end
+
+  def update(metadata)
+    metadata = metadata.sort_by { |row| row[Column::DATABASE_TYPE] }
     ::CSV.open @metadata_path, 'w' do |csv|
       metadata.each { |row| csv << row }
     end
-
-    logger.debug("metadata updated", :metadata => metadata)
+    logger.trace("metadata updated", :metadata => metadata)
   end
 
   def get_all
```

Comment:
I see the `get_all` -> transform -> `update(metadata)` pattern a lot.
Maybe we can instead introduce a helper:

```ruby
def update_metadata(&block)
   metadata = yield get_all
   update(metadata)
end
```

So that we can then implement, for example:

```ruby
def reset_md5(database_type)
  update_metadata do |metadata|
    metadata.map do |row|
      row[Column::GZ_MD5] = ""  if row[Column::DATABASE_TYPE].eql?(database_type)
      row
    end
  end
end
```